### PR TITLE
Hods 154

### PR DESCRIPTION
**Pull request by ajchilts**
Dependant on https://jira.tools.tax.service.gov.uk/browse/HODS-153 

Test using scope: read:organisations-matching-ho-ssp

Endpoints to hit;

{{OrgsMatchingBaseURL}}/self-assessment/4ecd50cd-d3ad-4985-93ef-bdf9940a07ec

{{OrgsMatchingBaseURL}}/corporation-tax/4ecd50cd-d3ad-4985-93ef-bdf9940a07ec

SM port: 9657

### CODE REVIEW

File: app/uk/gov/hmrc/organisationsmatchingapi/controllers/MatchedController.scala [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  with ErrorHandling
+  with PrivilegedAuthentication {
+
+  def matchedOrganisationCt(matchId: UUID): Action[AnyContent] = Action.async { implicit request =>
+    authenticate(scopeService.getAllScopes, matchId.toString) { authScopes =>
+      val correlationId = validateCorrelationId(request)
+      matchedService.fetchCt(matchId) map { cacheData =>
+
+        val exclude  = Some(List("details-self-assessment"))
+        val selfLink = HalLink("self", s"/organisations/matching/corporation-tax/$matchId")
```
Comment (parkeradam) [inline app/uk/gov/hmrc/organisationsmatchingapi/controllers/MatchedController.scala:53]: worth pulling out the interpolated string into a variable as its used both here and line 63.

File: app/uk/gov/hmrc/organisationsmatchingapi/controllers/MatchedController.scala [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    } recover recoveryWithAudit(maybeCorrelationId(request), request.body.toString, s"/organisations/matching/corporation-tax?matchId=$matchId")
+  }
+
+  def matchedOrganisationSa(matchId: UUID) : Action[AnyContent] = Action.async { implicit request =>
+    authenticate(scopeService.getAllScopes, matchId.toString) { authScopes =>
+      val correlationId = validateCorrelationId(request)
+      matchedService.fetchSa(matchId) map { cacheData =>
+
+        val exclude  = Some(List("details-corporation-tax"))
+        val selfLink = HalLink("self", s"/organisations/matching/self-assessment/$matchId")
```
Comment (parkeradam) [inline app/uk/gov/hmrc/organisationsmatchingapi/controllers/MatchedController.scala:72]: same as above

File: test/component/uk/gov/hmrc/organisationsmatchingapi/controllers/stubs/BaseSpec.scala [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  protected def requestHeaders(acceptHeader: (String, String) = acceptHeaderP1) =
+    Map(CONTENT_TYPE -> JSON, AUTHORIZATION -> authToken, acceptHeader, correlationIdHeader)
+
+  protected def requestHeadersInvalid(acceptHeader: (String, String) = acceptHeaderP1) =
+    Map(CONTENT_TYPE -> JSON, AUTHORIZATION -> authToken, acceptHeader)
+
+  protected def requestHeadersMalformed(acceptHeader: (String, String) = acceptHeaderP1) =
+    Map(CONTENT_TYPE -> JSON, AUTHORIZATION -> authToken, acceptHeader, correlationIdHeaderMalformed)
+
+  protected def errorResponse(message: String) =
```
Comment (parkeradam) [inline test/component/uk/gov/hmrc/organisationsmatchingapi/controllers/stubs/BaseSpec.scala:72]: Might be better to rename to invalid RequestResponse as errorResponse is a general class for errors in the system while this is only doing 1.

**Review by parkeradam — APPROVED** [review]

File: app/uk/gov/hmrc/organisationsmatchingapi/controllers/BaseApiController.scala [hunk 4]
```diff
@@ -51,6 +56,71 @@ abstract class BaseApiController (cc: ControllerComponents) extends BackendContr
           "message" -> "The request body does not conform to the schema.",
           "errors" -> Json.toJson(errors.toList))))
 
+  private[controllers] def recoveryWithAudit(correlationId: Option[String], matchId: String, url: String)(
+    implicit request: RequestHeader,
+    auditHelper: AuditHelper): PartialFunction[Throwable, Result] = {
+    case _: MatchNotFoundException => {
+      auditHelper.auditApiFailure(correlationId, matchId, request, url, "Not Found")
```
Comment (owen-opencast) [inline app/uk/gov/hmrc/organisationsmatchingapi/controllers/BaseApiController.scala:63]: I don't think this is an ApiFailure - a MatchNotFound is still a successful response with no match.

File: conf/app_version_v1_0.routes [hunk 5]
```diff
@@ -1,5 +1,5 @@
-POST        /corporation-tax                uk.gov.hmrc.organisationsmatchingapi.controllers.MatchingController.matchOrganisationCt
-GET         /corporation-tax/:matchId       uk.gov.hmrc.organisationsmatchingapi.controllers.MatchingController.matchedOrganisationCt(matchId: java.util.UUID)
+POST        /matching/corporation-tax                uk.gov.hmrc.organisationsmatchingapi.controllers.MatchingController.matchOrganisationCt
```
Comment (owen-opencast) [inline conf/app_version_v1_0.routes:1]: I don't think the routes should include the /matching/ part - I think it's added by the framework / environment. See individuals-matching.

**Review by owen-opencast — COMMENTED** [review]
A few comments 🙂

**Review by owen-opencast — APPROVED** [review]

**Pull request merged**